### DISCUSSION TITLE
Thoughts about this game's approach to difficulty

### DISCUSSION THREAD

**Discussion by onpon4** [singularity/singularity#332] [General]
I first played this game many years ago (I wanna say around 2012), but I was trying it again recently at the higher difficulties, and I can't help but feel like the game's approach to difficulty is working against it (and that while this isn't a huge problem at lower difficulty levels, it makes the higher difficulty levels more tedious rather than more difficult).

Currently the approach is very simple: every base has a certain chance over whatever the period of time is of getting destroyed, and if a base gets destroyed, the player is penalized by further increasing the chance of bases being destroyed, creating a positive feedback loop. This forces you to take things slow and not, for example, constantly build bases over and over again to grow exponentially. At its core, I like that the game's design limits you in this way.

When it's implemented in this very simple way, thô, it's really not all that interactive. As a player, all you can do is choose where to build, and what to build, to attempt to balance the chance of base destruction with your CPU needs. In theory you can also put all your bases into sleep mode to wait for suspicion to decrease, but this only reduces the chance of base destruction, so in practice, there's always going to be some level of suspicion. This means that very consistently, the core gameplay loop ends up being a repetitive motion of: see that a base was destroyed, build an identical base wherever is best, rinse, repeat. And since basic maintenance costs continually increase thruout the game, you're limited in your options as regards what you can realistically choose to build, so while in theory you can tailor your options for risk, in practice, you have very little influence over the process.

I particularly was thinking of this because, due to the problem where the game randomly crashes (which unfortunately I wasn't able to solve by either upgrading or downgrading Pygame), I was saving a whole lot to compensate. Before long, I started to realize something: because base destruction is just a random chance occurrence, there *is* one thing that a player can do to influence outcomes: save scumming, i.e. repeatedly saving and then reloading whenever a bad outcome happens. And as it happens, I found that save scumming is *less* tedious than playing the game normally. That's really indicative of a serious design flaw in my opinion.

I've been thinking of ways this game could create the intended effect of creating tension by making you afraid of discovery, but in a manner that's interactive. A few ideas come to mind:

1. Rather than immediately destroying a base, discovery could instead cause a temporary probationary period where being rediscovered then causes base destruction. This would enable at least doing some amount of damage control by putting the base in sleep mode.
2. Discovery could not cause destruction of a base at all, but instead (in addition to increasing suspicion) cause an exponential increase in detection chance (for that base specifically) that slowly reduces over time. This would mean that carelessness would lead to a different kind of chain reaction where failing to control the situation by either putting the base to sleep or (if it's too fargone) destroying it could quickly lose you the game. One thing I like about this idea is it would be putting the player in control of the currently mandatory scorched earth policy.
3. Whether or not discovery destroys a base could be based on whether or not a base was discovered there recently (reusing the currently existing mechanic where the last two discovery locations have an increased discovery chance). That would effectively be similar to idea 1, but implemented in a different way.
4. Bases made up of multiple components could be protected from complete destruction by security items, or something, maybe causing only the security items to be destroyed. (This is the idea I like the least, but still an idea worth mentioning.)

I came up with idea 2 as I was writing out idea 1, but I think I really like that one personally. I like how it gives agency to the player and allows them to choose what level of risk to take, while encouraging the player to still tend toward a sort of scorched earth policy as in the current implementation (albeit likely not quite as extreme as always destroying any base that gets detected no matter what).

In any case, I think something ought to be done to address this design issue.

**Comment by Anrock** [comment]
Somewhat relevant idea on how to address the design flaw
https://github.com/singularity/singularity/issues/20#issuecomment-601202725

**Comment by nthykier** [comment]
Hi

Thanks for the input. In general, I agree with the "lack of agency" problem for the player and thanks for your proposals for how one might change that.

On the crashing problem, this is most likely pygame.  I have had success with pygame 1.9.6 and pygame-ce 2.3.0 (the new community edition/fork). Allegedly pygame 2.5.0 also fixes the problem, but I did not test it myself (pygame/pygame#3128).


With this, we have the proposals:

1. Some form of action the player can take that can lowers suspicion actively (based on Anrock's remark / #20).
   - Alternative variants, actions to counter the detection rate of a group.  It still makes the suspicion passive, but you have an active stake in the problem rather than "huddle up and hope".
2. Some form of action the player can take when a base is discovered/about to discovered (based on the OP) where the player picks a consequence (base loss being one option).
    - If this method provides a way to prevents or strongly neuter the suspicion build up, then it might alleviate the need for 1).
    - Personally, I thought of a variant where detection is an analysis that builds up over time.  Suspicion raising when the longer the analysis has run if it finds evidence (base destruction maybe being a "cut your losses" strategy). Initially, the AI realizes the analysis late but with research, it finds them earlier giving the player more leeway.

Obviously, we also need a "counter balance" here.  The game is already "easy but tedious" and both of the mechanics proposed so far would probably make it even easier as you can either actively lower your "death counter" or "stall it from raising" as fast (which in the steady state is basically the same thing).

Are any of you interesting in coming up with a concrete design proposal for any of these mechanics? :)  (Think concrete researches, choices and their consequences, CPU allocation to effect - that kind of thing) 

I could very much do with a co-developer/game designer on this game.  If you can code as well, great - if coding is not your cop of tea, having concrete ideas for balancing with concrete numbering is fine too. :)

**Reply by Anrock** [reply]
Sorry, I'm afraid I don't have time to do dedicated balancing effort with concrete numbers nor do I have a skill or experience to do so. So the best effort here I can offer is brainstorming via meager comments. I can code but zero experience with python.

Back to business. I'm not sure if I get the difference between 1) and 2) you mentioned. But the point of not giving the player an ability to spam "lower suspicion" action is very valid. After reading your comment I think we should kinda combine both mentioned approaches. I'll try to formulate the systems. Maybe it's a bit overboard but I'll spill it all to be filtered.

---
# Proto intruder detection system
Right from the start AI is able to notice "probing" of its bases but in **non-systematic**, **unreliable** and **vague** way, maybe even with **false positives**.

The player will see a popup messages with vague fluff text that hints that some specific base is being scanned/surveyd, something like "you have noticed a port scan spike on rent server 55" or "motion sensors on warehouse 42 are picking something". The messages may be or may be not triggered by actual actions of hostile party (**unrealible**, **false positive**) and kept **vague** to leave player wondering if it was a stray cat triggering motion sensor or it's the FBI investigating the place. There should be no immediate actions attached to the message boxes or indicators on the main screen (**non-systematic**).

Next there is two options:
1) Leave the player in the dark on what to do until his first base is destroyed
2) Immediately show one-time hint with AI inner dialogue on how countermeasures are needed

# Intruder Detection System (IDS)
After base destruction or after first detection hint the research item is unlocked. It could also be made a prerequisite for existing suspicion indicators research item (sorry, haven't played in a loooong while so my research tree memory is very rusty).

Completing the research will remove the vague messages from proto system and introduce new upkeep CPU allocation for intrusion detection system.

## IDS upkeep
Upkeep CPU should be scaled with number of bases. 
It could be scaled non-linearly:
* Bigger bases take more CPU cycles (could we reuse current CPU base upkeep?)
* Be affected by installed security items (smart video cameras detect movement so less frames needs to be analyzed for example)?
* By additional research (smarter smart cameras i dunno)?
* Maybe even make it optional per-base: no cameras so no CPU upkeep but no IDS either

If upkeep CPU time is fully satisfied then when the base is detected by hostile party the damage control event is triggered instead of base destruction. If the upkeep is not satisfied then there is a dice roll for damage control event scaled with satisfied upkeep percentage.

## Damage control events
Basically upgraded popup messages from proto system but now clear, realible, without false positives and attached actions.

I think @onpon4 already listed some good options:
1) Put the base to sleep and hope for the best (is it pure random or we add some advocacy here like displaying chances of being left alone or something like this? previous detection indicator or just straight up lock this option if there was recent detection already?)
2) Create a loud distraction for short-term win and long-term loss (could be locked behind research, more on that later)
3) I would like to add another option here: soft destruction. Same as destruction - you lose the base but you also get some fraction of spent resources. "Oh, my warehouse is detected! Better sell those servers on ebay and get a refund for remaining monthly rent"

# Active distractions
Basically whatever I've proposed before. New research item(s) which are unlocked after getting global sus meter and IDS are done.

The item description is something like "techniques of generating false data, breaching into inner networks of hostile parties and planting sleeping trojans and other malicious stuff there". It should be pretty expensive item as it gives the player powerful tool to evade suspicion. Maybe even it could be splitted into per-faction research items with varying prices. Makes sense I think and also gives player an option to invest granularly instead of one big bite.

Anyway, after completing this research the new one repeatable item is unlocked (or many with one per-faction) which on completion will lower your suspicion as a result of everyone being distracted by whatever strawman you created.

To prevent player from spamming it should either be put under cooldown, have exponential costs when spamming (gotta falsify more and better proofs to keep attention) or diminishing returns (who will pay attention to fifteenth perpetual engine developed in this month?)

P.S. I don't remember if it's already in the game or not but it would be funny if those events were triggered randomly with same effects without player actions. Gives a bit of liveliness to the world and could be used to introduce the research item:
1) Have IDS and global sus meter unlocked
2) Random event is triggered
3) AI be like "hey, they got distracted. i could use this to my advantage if only i can make up some stuff and hype it up **research unlocked**"

**Comment by onpon4** [comment]
I wouldn't mind working on balancing a system, and could even do a little coding if necessary. I rather like this game conceptually, so I'd love to see its mechanics polished to make it a more engaging experience.

Regarding how to solve the problem of the game actually being easy, that's a bit challenging because of how simple the game is in combination with the fact that all bad outcomes are a result of random events, while most good outcomes are a result of systematic long-term processes. For example, if you've got a 10 CPU base researching a 500 CPU upgrade and it gets destroyed or disabled after two days, your only setback is that you're not actively doing anything; you'll just continue the rest of the 480 CPU toward the research on another base.

To be clear, that's the right design for research (having the player lose progress randomly would just be frustrating), but this game is in a unique position of choosing what to research and what bases to build being the *only* interaction. In your typical RTS, for example, you're making the same sorts of research and resource management decisions, but you're also organizing defenses against threats to them. In that situation, any random numbers involved in the destruction of your base are subtle enough that it wouldn't make sense to reload a save and then do the same thing hoping for a different outcome.

With that in mind, as far as making the game more challenging goes, there would need to be a shift from just random number decisions to *interactive events*. Similar to how you might, say, be able to mount a defense against a rush of tanks coming your way in a typical RTS, Endgame: Singularity will need to have its design changed to create threats that the player can respond to in meaningful ways, and which save scumming is unlikely to affect the creation of.

For example, taking a page out of typical RTS games, perhaps discovery chance could be changed to *investigation* chance. At the time that currently leads to discovery, the respective group would begin an investigation in the background. The player wouldn't be notified right away, but instead, a discovery event would be scheduled, where the "discovery" part is analogous to the actual battle in a typical RTS. Discovery, then, should be a very interactive process with little in the way of new random number usage.

It's likely that any such shift will make the game seem on paper to be easier than it currently is *initially*, but if a good, interactive system is worked out, it can then be modified to add new sources of difficulty afterwards.

I'll try to organize my thoughts into something more concrete later on; my thoughts are a little disorganized at the moment. 😆

**Comment by onpon4** [comment]
Sorry it took us so long to follow up on this. I think we won't be able to really help much after all; too much on our plate at the moment. That said, I think what I would do to solve the problem is simply to replace base destruction with a secondary, base-specific suspicion mechanic which can't cause loss of the game on its own, but exponentially increases the chance of the base being discovered each time a discovery happens. With discovery still raising global suspicion, of course. That would make discovery more like a threat that can be reacted to in various ways, rather than just something that happens to you and forces you to start rebuilding. A very simple change, but I think it would make a world of difference.
